src/deep-log/lib/Driver/RUM/AbstractEvent.js

```diff
@@ -0,0 +1,186 @@
+import Joi from 'joi';
+import {UnknownRumEventException} from '../Exception/UnknownRumEventException';
+import {FrameworkEvent} from './FrameworkEvent';
+
+/**
+ * Abstract RUM event
+ */
+export class AbstractEvent extends Core.OOP.Interface {
+  /**
+   * @param {Object} kernel
+   * @param {Object} rawData
+   */
+  constructor(kernel, rawData) {
+    super(['toJSON', 'getValidationSchema', 'getEventLevel']);
+
+    this._kernel = kernel;
+    this._rawData = rawData;
+    this._data = this._enrichWithContextData(rawData);
+
+    this._validationError = null;
+  }
+
+  /**
+   * @returns {String}
+   */
+  static get FRAMEWORK_EVENT_LEVEL() {
+    return 'Framework';
+  }
+
+  /**
+   * @returns {String}
+   */
+  static get BACKEND_CONTEXT() {
+    return 'Backend';
+  }
+
+  /**
+   * @returns {String}
+   */
+  static get FRONTEND_CONTEXT() {
+    return 'Frontend';
+  }
+
+  /**
+   * @returns {String[]}
+   */
+  static get CONTEXTS() {
+    return [
+      AbstractEvent.BACKEND_CONTEXT,
+      AbstractEvent.FRONTEND_CONTEXT,
+    ];
+  }
+
+  /**
+   * @returns {String[]}
+   */
+  static get SERVICES() {
+    return [
+      'deep-asset',
+      'deep-cache',
+      'deep-core',
+      'deep-db',
+      'deep-event',
+      'deep-fs',
+      'deep-kernel',
+      'deep-log',
+      'deep-notification',
+      'deep-resource',
+      'deep-security',
+      'deep-validation',
+    ];
+  }
+
+  /**
+   * @returns {String[]}
+   */
+  static get RESOURCE_TYPES() {
+    return [
+      'Browser',
+      'Lambda',
+      'S3',
+      'DynamoDB',
+      'CloudFront',
+      'InMemory',
+      'LocalStorage',
+      'Redis',
+      'S3FS',
+      'Cognito',
+    ];
+  }
+
+  /**
+   * @param {Object} kernel
+   * @param {Object} rawData
+   * @returns {AbstractEvent}
+   */
+  static create(kernel, rawData) {
+    let event = null;
+
+    // @note - For the time being event type is guessed by event.service
+    // (it'll be changed once we'll have other types of events)
+    if (rawData.service && AbstractEvent.SERVICES.indexOf(rawData.service) !== -1) {
+      event = new FrameworkEvent(kernel, rawData);
+    } else {
+      throw new UnknownRumEventException(rawData);
+    }
+
+    return event;
+  }
+
+  /**
+   * @returns {Boolean}
+   */
+  isValid() {
+    let result = this.validate();
+
+    return result.error ? false : true;
+  }
+
+  /**
+   * @returns {Object}
+   */
+  validate() {
+    let result = Joi.validate(this._rawData, this.getValidationSchema(), {
+      stripUnknown: true,
+      convert: true,
+      abortEarly: false,
+    });
+
+    if (result.error) {
+      this._validationError = result.error;
+    } else {
+      this._data = result.value;
+    }
+
+    return result;
+  }
+
+  /**
+   * @returns {Object|null}
+   */
+  get validationError() {
+    return this._validationError;
+  }
+
+  /**
+   * @param {Object} event
+   * @returns {Object}
+   * @private
+   */
+  _enrichWithContextData(event) {
```

Comment:
![Codacy](https://www.codacy.com/versioned/images/favicon.png) Issue found: [This function's cyclomatic complexity is too high. (8)](https://www.codacy.com/app/MitocGroup/deep-framework/file/2044864354/issues/source?bid=2895475&fileBranchId=3129256#l158)